File: Smile.Client/src/app/services/group.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { map } from 'rxjs/operators';
import { environment } from 'src/environments/environment';
import { Group } from '../models/domain/group/group';
import { PaginatedResult } from '../models/helpers/pagination';
import { GroupsRequest } from '../resolvers/requests/groups-request';
import { GroupsResponse } from '../resolvers/responses/groups-response';
import { UserGroupsResponse } from '../resolvers/responses/user-groups-response';

@Injectable({
  providedIn: 'root'
})
export class GroupService {
  private readonly groupApiUrl = environment.apiUrl + 'group/';

  constructor(private httpClient: HttpClient) { }

  public getGroup(groupId: string) {
    return this.httpClient.get<Group>(this.groupApiUrl, { params: { groupId } });
  }

  public fetchGroups(groupsRequest: GroupsRequest) {
    const paginatedResult: PaginatedResult<GroupsResponse> = new PaginatedResult<GroupsResponse>();

    let httpParams = new HttpParams();

    httpParams = httpParams.append('pageNumber', groupsRequest.pageNumber.toString());
    httpParams = httpParams.append('pageSize', groupsRequest.pageSize.toString());

    if (groupsRequest.name) {
      httpParams = httpParams.append('name', groupsRequest.name);
    }

    if (groupsRequest.accessStatus) {
      httpParams = httpParams.append('accessStatus', groupsRequest.accessStatus.toString());
    }

    if (groupsRequest.joinStatus) {
      httpParams = httpParams.append('joinStatus', groupsRequest.joinStatus.toString());
    }

    if (groupsRequest.sortType) {
      httpParams = httpParams.append('sortType', groupsRequest.sortType.toString());
    }

    if (groupsRequest.isInvited) {
      httpParams = httpParams.append('isInvited', groupsRequest.isInvited.toString());
    }

    return this.httpClient.get<GroupsResponse>(this.groupApiUrl + 'filter', { observe: 'response', params: httpParams })
      .pipe(
        map(response => {
          paginatedResult.result = response.body;
          if (response.headers.get('Pagination')) {
            paginatedResult.pagination = JSON.parse(response.headers.get('Pagination'));
          }

          return paginatedResult;
        })
      );
  }

  public createGroup(request: any, image?: File) {
    const formData = new FormData();

    formData.append('name', request.name);

    if (request.description) {
      formData.append('description', request.description);
    }

    formData.append('isPrivate', request.isPrivate.toString());

    if (image) {
      formData.append('image', image);
    }

    if (request.joinCode) {
      formData.append('joinCode', request.joinCode);
    }

    formData.append('inviteMemberPermission', request.inviteMemberPermission.toString());
    formData.append('removeMemberPermission', request.removeMemberPermission.toString());

    return this.httpClient.post(this.groupApiUrl + 'create', formData, { observe: 'response' });
  }

  public joinGroup(groupId: string, joinCode?: string) {
    return this.httpClient.post(this.groupApiUrl + 'join', { groupId, joinCode });
  }

  public fetchUserGroups() {
    return this.httpClient.get<UserGroupsResponse>(this.groupApiUrl + 'user/groups');
  }
}

File: Smile.Infrastructure.Shared/Services/GroupService.cs
using Microsoft.AspNetCore.Http;
using Smile.Core.Common.Enums.Permissions;
using Smile.Core.Domain.Data;
using System.Linq;
using System.Threading.Tasks;
using Smile.Core.Application.Builders;
using Smile.Core.Application.Exceptions;
using Smile.Core.Application.Logic.Requests.Query.Group;
using Smile.Core.Application.Results;
using Smile.Core.Application.Services;
using Smile.Core.Application.Services.ReadOnly;
using Smile.Core.Domain.Entities.Group;
using Smile.Core.Domain.Data.Models;

namespace Smile.Infrastructure.Shared.Services
{
    public class GroupService : IGroupService
    {
        private readonly IDatabase database;
        private readonly IFilesManager filesManager;
        private readonly IReadOnlyProfileService profileService;

        public GroupService(IDatabase database, IFilesManager filesManager, IReadOnlyProfileService profileService)
        {
            this.database = database;
            this.filesManager = filesManager;
            this.profileService = profileService;
        }

        public async Task<Group> GetGroup(string groupId)
        {
            var group = await database.GroupRepository.Get(groupId) ??
                        throw new EntityNotFoundException("Group not found");

            var currentUser = await profileService.GetCurrentUser();
            var currentUserGroups =
                currentUser.Groups.Concat(currentUser.GroupMembers.Where(m => m.IsAccepted).Select(m => m.Group));

            return group.AdminId == currentUser.Id ||
                   group.GroupMembers.Any(m => m.UserId == currentUser.Id && m.IsAccepted) || currentUser.IsAdmin()
                ? group
                : throw new NoPermissionsException("You are not member of this group");
        }

        public async Task<IPagedList<Group>> FetchGroups(FetchGroupsPaginationRequest paginationRequest)
            => await database.GroupRepository.GetFilteredGroups(paginationRequest,
                (paginationRequest.PageNumber, paginationRequest.PageSize));

        public async Task<Group> CreateGroup(string name, string description, bool isPrivate, IFormFile image,
            string joinCode,
            InviteMemberPermission inviteMemberPermission, RemoveMemberPermission removeMemberPermission)
        {
            if (await GroupExists(name))
                throw new DuplicateException("Group already exists");

            var currentUser = await profileService.GetCurrentUser();

            var group = new GroupBuilder()
                .SetName(name)
                .WithDescription(description)
                .IsPrivate(isPrivate)
                .WithJoinCode(joinCode)
                .WithPermissions(inviteMemberPermission, removeMemberPermission)
                .Build();

            if (image != null)
            {
                var uploadedImage = await filesManager.Upload(image, $"groups/{group.Id}");
                group.SetImage(uploadedImage?.Path);

                database.FileRepository.AddFile(uploadedImage?.Path);
            }

            currentUser.Groups.Add(group);

            return await database.Complete() ? group : null;
        }

        public async Task<GroupMember> JoinGroup(string groupId, string joinCode = null)
        {
            var group = await database.GroupRepository.Get(groupId) ??
                        throw new EntityNotFoundException("Group not found");
            var currentUser = await profileService.GetCurrentUser();

            if (group.AdminId == currentUser.Id || group.GroupMembers.Any(m => m.UserId == currentUser.Id))
                throw new DuplicateException("You are member of this group");

            return await CreateMember(group, currentUser.Id, joinCode);
        }

        public async Task<UserGroupsResult> FetchUserGroups()
        {
            var currentUser = await profileService.GetCurrentUser();

            var ownGroups = currentUser.Groups.OrderByDescending(g => g.DateCreated).ToList();
            var myGroups = currentUser.GroupMembers.Where(m => m.IsAccepted).Select(m => m.Group)
                .OrderByDescending(g => g.DateCreated).ToList();

            return new UserGroupsResult(ownGroups, myGroups);
        }

        #region private

        private async Task<GroupMember> CreateMember(Group group, string userId, string joinCode)
        {
            var member = GroupMember.Create(userId, group.Id);

            if (group.IsPrivate && !string.IsNullOrEmpty(group.JoinCode) && group.JoinCode != joinCode)
                throw new NoPermissionsException("Invalid join code");
            else if (!group.IsPrivate ||
                     (group.IsPrivate && !string.IsNullOrEmpty(group.JoinCode) && group.JoinCode == joinCode))
                member.Accept();
            else
            {
                var groupInvite = GroupInvite.Create(userId, group.Id, isInvited: false);
                group.GroupInvites.Add(groupInvite);
            }

            database.GroupMemberRepository.Add(member);

            return await database.Complete() ? member : null;
        }

        private async Task<bool> GroupExists(string groupName) =>
            await database.GroupRepository.Find(g => g.Name.ToLower() == groupName.ToLower()) != null;

        #endregion
    }
}